File: 4 Kyu/4 Kyu - Matrix determinant.cpp
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

int64_t determinant (vector<vector<int64_t>> mat) {

    int64_t det = 0;

    switch (mat.size()) {
        case 0 : det = 0; break;
        case 1 : det = mat[0][0]; break;
        case 2 : det = mat[0][0] * mat[1][1] - mat[0][1] * mat[1][0]; break;
        default : {

            short sign = 1;

            for (int dig = 0; dig < mat.size(); ++dig) {
                vector<vector<int64_t>> sub;

                for (int i = 0; i < mat.size(); ++i) {
                    vector<int64_t> v;
                    for (int j = 0; j < mat.size(); ++j) {
                        if (i != 0 && j != dig)
                            v.push_back(mat[i][j]);
                    }

                    if (i != 0) sub.push_back(v);
                }

                det += sign * mat[0][dig] * determinant (sub);
                sign *= -1;
            }
        } break;
    }

    return det;
}

int main () {

    //Test();
    vector<vector<int64_t>> matrix {
      {2,5,3},
      {1,-2,-1},
      {1,3,4}
    };
    vector<vector<int64_t>> tab {
      {1,3},
      {2,5}};

}

File: 4 Kyu/4 Kyu - Matrix determinant.rs

fn determinant(mx: &[Vec<i64>]) -> i64 {

    match mx.len() {
        1 => return mx[0][0],
        2 => return mx[0][0] * mx[1][1] - mx[0][1] * mx[1][0],
        _ => {
            let mut det:i64 = 0;
            let mut sign:i64 = -1;

            for dig in 0..mx.len() {
                let mut nxt:Vec<Vec<i64>> = Vec::new();
                sign *= -1;

                for i in 0..mx.len() {
                    let mut ve:Vec<i64> = Vec::new();

                    for j in 0..mx.len() {

                        if i != 0 && j != dig { ve.push (mx[i][j]); }
                    }

                    if i != 0 { nxt.push (ve) }
                }
                det += sign * mx[0][dig] * determinant (&nxt);
            }
            det
        },
    }

}
